flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -489,16 +486,15 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
 	//  Utilities
 	// ------------------------------------------------------------------------
 
-	private void assignTopicPartitions(List<KafkaTopicPartition> kafkaTopicPartitions) {
-		subscribedPartitions = new ArrayList<>();
-
+	private void assignTopicPartitions() {
 		if (restoreToOffset != null) {
-			for (KafkaTopicPartition kafkaTopicPartition : kafkaTopicPartitions) {
-				if (restoreToOffset.containsKey(kafkaTopicPartition)) {
-					subscribedPartitions.add(kafkaTopicPartition);
-				}
+			subscribedPartitions = new ArrayList<>(restoreToOffset.size());
+			for (Map.Entry<KafkaTopicPartition, Long> restoredPartitionState : restoreToOffset.entrySet()) {
+				subscribedPartitions.add(restoredPartitionState.getKey());
```

Comment:
(note about your comment --->) subscribed partitions will always be completely identical to the restored state, if there is any.

I should actually just change this to `subscribedPartitions = new ArrayList<>(restoreToOffset.keySet);`, to be more clearer.